Sources/AblyChat/Rooms.swift

```diff
@@ -1,5 +1,64 @@
+@preconcurrency import Ably
+
 public protocol Rooms: AnyObject, Sendable {
     func get(roomID: String, options: RoomOptions) throws -> any Room
     func release(roomID: String) async throws
     var clientOptions: ClientOptions { get }
 }
+
+internal final class DefaultRooms: Rooms {
+    private let realtime: ARTRealtimeProtocol
+    private let storage = Storage()
+
+    internal init(realtime: ARTRealtimeProtocol) {
+        self.realtime = realtime
+    }
+
+    /// Thread-safe storage of the set of rooms.
+    private class Storage: @unchecked Sendable {
```

Comment:
Why you'd prefer using `unchecked`+`let storage`+`Lock` here instead of just using `actor`?

Reply:
I don't know what the right thing to do is, to be honest. The thing with using actors is that to mutate an actor’s state from outside the actor, you have to await the result of the mutation. That would mean, in this example, that we would have to make the `Rooms` protocol’s `get(roomID:options:)` method `async`.

So, yes, using actors makes the implementation nicer and gives you a lot of support from the language, but at the cost of getting some weird-looking APIs. I wondered the same thing when I was working on #30, in which I used an actor and ended up having to, for example, make the `Messages.subscribe` method `async`, which looked weird. So I thought that in this PR I’d try something different to avoid having to change the public API 🤷

Reply:
IMHO `get(roomID:options:)` makes sense being `async` from the POV of the API user, `subscribe` not so much, but at least this is the async framework build on top of another async framework with full language support of concurrency, so marking everything as `async` kinda makes sense and shouldn't be shocking for the user. But I'm leaving it for yourself to decide. also @umair-ably wdyt?

Reply:
If as a team we decide that we're going to use actors and accept `async` "polluting" (put in quotes because I guess it's a subjective thing) a large chunk of our public API, I am happy to go along with that.

Reply:
Yeah, if we decide to fully support async/concurrency, then why trying to dodge it, right? I vote for using async everywhere.

Reply:
Also it will let us not to use `unchecked` which is eye irritating.

Reply:
I won't go as far as saying `everything` should be async 😅 but I do think it makes sense for both get rooms and subscribing to a message. It does also seem safer to not operate outside of compiler checks as `unchecked` does

Reply:
We can try with the actors approach and see how the API looks. If we're not happy with it later, though, we might have to go back on it.

It's worth pointing out that there's nothing _wrong_ with an approach that uses locks; it's just that the compiler can't reason about it.

In Swift 6 / iOS 18+, you can use the `Mutex` type, which lets you synchronously manage mutable state in a way that doesn't require you to write `@unchecked Sendable`. I’m not sure it's possible to write an equivalent without language support.

```swift
struct Foo: Sendable {
    var bar: String
}

@available(macOS 15.0, *)
final class MyThing: Sendable {
    let mutex = Mutex<[Foo]>([.init(bar: "hello")])

    func doThing() {
        mutex.withLock { value in
            value.append(.init(bar: "goodbye"))
        }
    }
}
```

Reply:
Strange that swift still doesn't have a property wrapper equal to `atomic` keyword in objective-c.

Reply:
You could, I think, write a property wrapper that implements the `@atomic` behaviour, but it's not really what you want. In the example you've given, if you were to call `doThing()` from two separate threads, it's possible for both threads to read `"hello"` before either one writes, meaning that the final value of `foo.bar` would be `"hello goodbye"` instead of the intended `"hello goodbye goodbye"`.

Reply:
(what you want is a lock that wraps the (read + write) transaction, which a property wrapper couldn't do)

Reply:
Interesting, thanks for pointing this out 👍

Reply:
> We can try with the actors approach and see how the API looks. If we're not happy with it later, though, we might have to go back on it.

👌 

I think this is preferred if anything. We're still early on in this project so we can prioritise speed over perfection, just so we get to uncover the real issues down the road (if any) a lot quicker. None of these decisions are so huge that we can't turn back from them either

Reply:
I’ve switched it to use actors, and I’ve put a summary of this conversation / decision in the commit message.